# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance','NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()
df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  
0                  1        101348.88       1  
1                  1        112542.58       0  
2                  0        113931.57       1  
3                  0         93826.63       0  
4                  1         79084.10       0  
...              ...              ...     ...  
9995               0         96270.64       0  
9996               1        101699.77       0  
9997               1         42085.58       1  
9998               0         92888.52       1  
9999               0         38190.78       0  

[10000 rows x 10 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Geography        10000 non-null  object 
 1   Gender           10000 non-null  object 
 2   Age              10000 non-null  int64  
 3   Tenure           10000 non-null  int64  
 4   Balance          10000 non-null  float64
 5   NumOfProducts    10000 non-null  int64  
 6   HasCrCard        10000 non-null  int64  
 7   IsActiveMember   10000 non-null  int64  
 8   EstimatedSalary  10000 non-null  float64
 9   Exited           10000 non-null  int64  
dtypes: float64(2), int64(6), object(2)
memory usage: 781.4+ KB


## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [ ]:
df

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      42       2       0.00              1          1               1   
1      41       1   83807.86              1          0               1   
2      42       8  159660.80              3          1               0   
3      39       1       0.00              2          0               0   
4      43       2  125510.82              1          1               1   
...   ...     ...        ...            ...        ...             ...   
9995   39       5       0.00              2          1               0   
9996   35      10   57369.61              1          1               1   
9997   36       7       0.00              1          0               1   
9998   42       3   75075.31              2          1               0   
9999   28       4  130142.79              1          1               0   

      EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0           101348.88       1                 1                  0   
1           112542.58       0                 0                  0   
2           113931.57       1                 1                  0   
3            93826.63       0                 1                  0   
4            79084.10       0                 0                  0   
...               ...     ...               ...                ...   
9995         96270.64       0                 1                  0   
9996        101699.77       0                 1                  0   
9997         42085.58       1                 1                  0   
9998         92888.52       1                 0                  1   
9999         38190.78       0                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop("Exited", axis=1)
y = df["Exited"]


In [ ]:
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [ ]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

In [ ]:
X_transform

Age  Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     0.324324     0.2  0.000000       0.000000        1.0             1.0   
1     0.310811     0.1  0.334031       0.000000        0.0             1.0   
2     0.324324     0.8  0.636357       0.666667        1.0             0.0   
3     0.283784     0.1  0.000000       0.333333        0.0             0.0   
4     0.337838     0.2  0.500246       0.000000        1.0             1.0   
...        ...     ...       ...            ...        ...             ...   
9995  0.283784     0.5  0.000000       0.333333        1.0             0.0   
9996  0.229730     1.0  0.228657       0.000000        1.0             1.0   
9997  0.243243     0.7  0.000000       0.000000        0.0             1.0   
9998  0.324324     0.3  0.299226       0.333333        1.0             0.0   
9999  0.135135     0.4  0.518708       0.000000        1.0             0.0   

      EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0            0.506735               1.0                0.0              0.0   
1            0.562709               0.0                0.0              1.0   
2            0.569654               1.0                0.0              0.0   
3            0.469120               1.0                0.0              0.0   
4            0.395400               0.0                0.0              1.0   
...               ...               ...                ...              ...   
9995         0.481341               1.0                0.0              0.0   
9996         0.508490               1.0                0.0              0.0   
9997         0.210390               1.0                0.0              0.0   
9998         0.464429               0.0                1.0              0.0   
9999         0.190914               1.0                0.0              0.0   

      Gender_Female  Gender_Male  
0               1.0          0.0  
1               1.0          0.0  
2               1.0          0.0  
3               1.0          0.0  
4               1.0          0.0  
...             ...          ...  
9995            0.0          1.0  
9996            0.0          1.0  
9997            1.0          0.0  
9998            0.0          1.0  
9999            1.0          0.0  

[10000 rows x 12 columns]

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

Model pertama yang digunakan adalah Decision Tree. Decision Tree adalah model yang digunakan untuk pengambilan keputusan di mana setiap cabang dari pohon merepresentasikan keputusan, dan setiap daun mewakili hasil dari keputusan tersebut. Dalam konteks DecisionTreeClassifier, model ini digunakan untuk klasifikasi, di mana tujuannya adalah memisahkan data ke dalam kelas atau kategori yang berbeda. Keunggulan utama dari model Decision Tree adalah kemampuannya untuk melakukan parameter tuning yang cermat. Dengan menggunakan GridSearchCV, model dapat secara otomatis mencari kombinasi hyperparameter terbaik yang meningkatkan performa model dalam memprediksi data. Hasil dari parameter tuning membantu meningkatkan akurasi dan kinerja model secara keseluruhan, memungkinkan penggunaan model ini secara optimal dalam tugas klasifikasi. Dengan demikian, parameter tuning adalah langkah kunci dalam menyempurnakan model Decision Tree untuk mencapai hasil terbaik dalam pemrosesan data dan prediksi.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.tree import DecisionTreeClassifier # Import class DecisionTreeClassifier dari modul sklearn.tree

#START CODE
model1 =  DecisionTreeClassifier() # Inisialisasi model DecisionTreeClassifier tanpa hyperparameter
params = {
    'criterion': ['gini', 'entropy'],  # Kriteria pemilihan fitur
    'max_depth': [None, 10, 20, 30],  # Maksimum kedalaman setiap pohon
    'min_samples_split': [2, 5, 10],  # Minimum sampel yang dibutuhkan untuk membagi node
    'min_samples_leaf': [1, 2, 4],  # Minimum sampel yang dibutuhkan di setiap leaf node
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV # Import class GridSearchCV dari modul sklearn.model_selection
grid_DT = GridSearchCV( # Inisialisasi objek GridSearchCV
             estimator= model1, # Model yang akan dioptimalkan
             param_grid= params,  # Daftar hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             n_jobs =10, # Core cpu yang digunakan
             cv =10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid_DT.fit(X_train,y_train) # Melakukan optimisasi hyperparameter menggunakan data training
grid_DT.best_params_ # Mendapatkan hyperparameter terbaik dari hasil optimisasi

{'criterion': 'entropy',
 'max_depth': 10,
 'min_samples_leaf': 2,
 'min_samples_split': 10}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid_DT.predict(X_test) # Melakukan prediksi menggunakan model yang telah ditest

print(classification_report(y_test,y_pred)) # Menampilkan classification report
print("")
print(confusion_matrix(y_test,y_pred)) # Menampilkan confusion matrix
print("")
print(accuracy_score(y_test, y_pred)) # Menampilkan nilai akurasi

              precision    recall  f1-score   support

           0       0.87      0.95      0.91      1983
           1       0.69      0.47      0.56       517

    accuracy                           0.85      2500
   macro avg       0.78      0.71      0.74      2500
weighted avg       0.84      0.85      0.84      2500


[[1874  109]
 [ 272  245]]

0.8476


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

Model kedua yang saya gunakan adalah Random Forest. Random Forest adalah sebuah model machine learning yang beroperasi dengan cara membuat beberapa pohon keputusan (decision trees) dan menggabungkan hasil prediksi dari setiap pohon untuk menghasilkan prediksi akhir. Pada setiap pohon keputusan, fitur-fitur yang digunakan untuk pemilihan keputusan dibuat secara acak, sehingga setiap pohon memiliki keberagaman dalam pembelajaran. Pada codingan yang diberikan, model Random Forest pertama kali dibuat tanpa melakukan penalaan terhadap hyperparameter, seperti jumlah pohon (n_estimators) dan kedalaman maksimum pohon (max_depth). Kemudian, dengan menggunakan GridSearchCV, kita mencari kombinasi terbaik dari hyperparameter untuk meningkatkan kinerja model. Random Forest diandalkan karena beberapa alasan, termasuk kemampuannya dalam mengatasi overfitting, ketahanannya terhadap noise, dan kemampuannya untuk menangani data dengan fitur yang banyak. Pemilihan fitur secara acak dan penggunaan beberapa pohon membuat model ini sangat efektif dalam berbagai jenis data dan tugas klasifikasi.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.ensemble import RandomForestClassifier # Import class RandomForestClassifier dari modul sklearn.tree

#START CODE
model2 = RandomForestClassifier() # Inisialisasi model RandomForestClassifier tanpa hyperparameter
params = {
    'n_estimators' : [100, 200, 300],  # Jumlah pohon dalam ensemble
    'max_depth': [3, 5, None], # Maksimum kedalaman setiap pohon
    'max_features' : ['sqrt', 'log2'], # Jumlah fitur yang dipertimbangkan untuk split
    'min_samples_split' : [2, 4, 6] # Minimum sampel yang dibutuhkan untuk membagi node
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV # Import class GridSearchCV dari modul sklearn.model_selection
grid_RF = GridSearchCV( # Inisialisasi objek GridSearchCV
             estimator= model2, # Model yang akan dioptimalkan
             param_grid= params, # Daftar hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             cv =10, # Core cpu yang digunakan
             n_jobs =10 # core cpu yang digunakan # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid_RF.fit(X_train,y_train) # Melakukan optimisasi hyperparameter menggunakan data training
grid_RF.best_params_ # Mendapatkan hyperparameter terbaik dari hasil optimisasi

{'max_depth': None,
 'max_features': 'log2',
 'min_samples_split': 6,
 'n_estimators': 200}

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid_RF.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.78      0.47      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.72      0.75      2500
weighted avg       0.85      0.86      0.85      2500


[[1914   69]
 [ 275  242]]

0.8624


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

Model ketiga yang digunakan dalam kode di atas adalah Support Vector Machine (SVM) melalui class SVC dari scikit-learn. SVM adalah model pembelajaran mesin yang sangat berguna untuk klasifikasi dan regresi. Pada kode tersebut, objek SVM dibuat dengan parameter default, kemudian dilakukan penyetelan parameter menggunakan Grid Search Cross Validation (GridSearchCV) untuk mencari kombinasi terbaik dari 'C' (cost of misclassification), 'kernel' (jenis kernel), dan 'gamma' (koefisien kernel) guna meningkatkan kinerja model SVM. Proses ini dilakukan dengan GridSearchCV menggunakan metode cross-validation, yang memastikan keakuratan dan ketahanan model. Keunggulan class SVC terletak pada kemampuannya menangani pemisah linier yang kompleks dalam data, sehingga efektif untuk masalah yang melibatkan pemisah yang rumit. Selain itu, SVM mampu menangani data dengan dimensi yang tinggi, menjadikannya relevan untuk berbagai jenis dataset dengan karakteristik yang berbeda. Dengan kombinasi parameter yang tepat, class SVC dapat memberikan hasil klasifikasi atau regresi yang sangat baik dalam berbagai aplikasi.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.svm import SVC # Import class SVC dari modul sklearn.svm

#START CODE
model3 = SVC() # Inisialisasi model Support Vector Classification (SVC) tanpa hyperparameter
params = {
    'C': [0.1, 1, 10],  # Parameter penalti
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],  # Jenis kernel
    'gamma': ['scale', 'auto', 0.1, 1],  # Koefisien kernel
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV # Import class GridSearchCV dari modul sklearn.model_selection
grid_SVC = GridSearchCV( # Inisialisasi objek GridSearchCV
             estimator= model3, # Model yang akan dioptimalkan
             param_grid= params, # Daftar hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             n_jobs = 10, # Core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid_SVC.fit(X_train,y_train) # Melakukan optimisasi hyperparameter menggunakan data training
grid_SVC.best_params_ # Mendapatkan hyperparameter terbaik dari hasil optimisasi

{'C': 10, 'gamma': 1, 'kernel': 'poly'}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid_SVC.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.98      0.92      1983
           1       0.82      0.44      0.57       517

    accuracy                           0.86      2500
   macro avg       0.85      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


[[1934   49]
 [ 289  228]]

0.8648


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya


Model terbaik adalah **Model 3, yaitu model SVM dengan class SVC** yang memiliki akurasi sebesar 0.8648. Alasannya karena akurasi yang tinggi menunjukkan bahwa model ini mampu melakukan prediksi dengan lebih baik dibandingkan Model 1 (Decision Tree) dengan akurasi 0.8476 dan Model 2 (Random Forest) dengan akurasi 0.8624. Selain akurasi, SVM (Support Vector Machine) juga efektif dalam menangani data yang kompleks dan memiliki kemampuan generalisasi yang baik. SVM mampu menangani kasus di mana hubungan antara fitur dan label bersifat non-linear. Oleh karena itu, keunggulan SVM class SVC dalam mengatasi kompleksitas dan mempertahankan tingkat akurasi yang tinggi membuatnya menjadi pilihan yang lebih baik di antara ketiga model tersebut.